src/libs/protoboard/protoboard_types.h

```diff
@@ -0,0 +1,97 @@
+
+/***************************************************************************
+ * Protoboard plugin template
+ * - General forward declarations
+ *
+ * Copyright 2019 Victor Mataré
+ ****************************************************************************/
+
+/*  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Library General Public License for more details.
+ *
+ *  Read the full text in the LICENSE.GPL file in the doc directory.
+ */
+
+#ifndef PROTOBOARD_TYPES_H_
+#define PROTOBOARD_TYPES_H_
+
+#include <boost/bimap.hpp>
+#include <memory>
+#include <unordered_map>
+
+namespace protoboard {
+
+class pb_convert;
+
+typedef std::unordered_map<std::string, std::shared_ptr<pb_convert>> pb_conversion_map;
+
+/**
+ * Helper structure to wrap a list of types into a single type.
+ */
+template <typename... Ts>
+struct type_list
+{
+};
+
+class BlackboardManager;
+
+template <class IfaceT>
+struct InterfaceMessageHandler;
+
+/**
+ * A compile-time constant bidirectional map that can be used to match blackboard interface enum
+ * values to ProtoBuf's enum values
+ * @tparam pbEnumT a ProtoBuf enum type
+ * @tparam bbEnumT a blackboard interface enum type
+ */
+template <class pbEnumT, class bbEnumT>
+class enum_map
+{
+private:
+	typedef boost::bimap<pbEnumT, bbEnumT> bimapT;
+
+public:
+	/**
+	 * constexpr constructor
+	 * @param init A curly-brace initializer list that defines the entire mapping
+	 */
+	constexpr enum_map(std::initializer_list<typename bimapT::value_type> init)
+	: list(init), map(list.begin(), list.end())
+	{
+	}
+
+	/**
+	 * @param v a ProtoBuf enum value
+	 * @return the mapped blackboard interface enum value
+	 */
+	constexpr bbEnumT
```

Comment:
Why not return by const reference?

Reply:
Well, if we can evaluate it all at compile time, why shouldn't we?

Reply:
The function can be `constexpr` and the return type `const bbEnumT &`, the former doesn't have anything to do with the latter. In particular, `constexpr` functions can return non-const values.

Reply:
Wow, I always thought of `constexpr` as a stronger `const`, but you're right, the two are vertical to each other. That said, it doesn't really matter here because this thing is just being used for enums. On the other hand, there's nothing preventing a user from putting other things in there, too...